[TS] Add complex support for more ops

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54541 [TS] Add complex support for more ops**

Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

anjali411 committed:
[TS] Add complex support for more ops

[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"




[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -924,12 +924,19 @@ RegisterOperators reg2({
-    DEFINE_SCALAR_SCALAR_BINARY_OP(
+    DEFINE_INT_COMPLEX_OP(aten::log, std::log(a) / std::log(b), complex),
+    DEFINE_FLOAT_COMPLEX_OP(aten::log, std::log(a) / std::log(b), complex),
+    DEFINE_SCALAR_BINARY_OP_WITH_COMPLEX(
```
```suggestion
    DEFINE_SCALAR_SCALAR_BINARY_OP_WITH_COMPLEX(
```
As discussed this should probably be `DEFINE_SCALAR_SCALAR_BINARY_OP_WITH_COMPLEX` to avoid naming conflicts with `aten::log` for tensors.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1057,6 +1076,18 @@ RegisterOperators reg2({
           push(stack, sum);
         },
         aliasAnalysisFromSchema()),
+    // Operator(
+    //     "aten::sum.complex(complex[] self) -> complex",
+    //     [](Stack* stack) {
+    //       c10::List<c10::complex<double>> l =
+    //       pop(stack).toComplexDoubleList();
+    //       c10::complex<double> sum = 0.0;
+    //       for (int i=0; i<l.size(); i++) {
+    //         sum += l[i];
+    //       }
+    //       push(stack, sum);
+    //     },
+    //     aliasAnalysisFromSchema()),
```
```suggestion
```

anjali411 replied:
no actually I need to fix that

SplitInfinity commented on torch/csrc/jit/frontend/sugared_value.cpp:
```diff
@@ -122,6 +122,8 @@ std::shared_ptr<SugaredValue> SimpleValue::attr(
            {"T", "prim"},
            {"ndim", "prim"},
            {"name", "prim"},
+           {"real", "prim"},
+           {"imag", "prim"},
```
```suggestion
           {"real", "aten"},
           {"imag", "aten"},
```

I found these in `native_functions.yaml`, let's try using them instead of defining our own:
```
- func: real(Tensor(a) self) -> Tensor(a)
  variants: function

- func: imag(Tensor(a) self) -> Tensor(a)
  variants: function
```

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -1067,6 +1086,22 @@ RegisterOperators reg(
            push(stack, a.is_cuda());
          },
          aliasAnalysisFromSchema()),
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA("prim::real(Tensor a) -> Tensor"),
+         [](Stack* stack) {
+           at::Tensor a;
+           pop(stack, a);
+           push(stack, at::real(a));
+         },
+         aliasAnalysisFromSchema()),
+     OperatorGenerator(
+         TORCH_SELECTIVE_SCHEMA("prim::imag(Tensor a) -> Tensor"),
+         [](Stack* stack) {
+           at::Tensor a;
+           pop(stack, a);
+           push(stack, at::imag(a));
+         },
+         aliasAnalysisFromSchema()),
```
Try removing these and using the ones from `native_functions.yaml` like I suggested above.

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


TODO: add tests


[ghstack-poisoned]

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1057,6 +1077,17 @@ RegisterOperators reg2({
+        [](Stack* stack) {
+          c10::List<c10::complex<double>> l = pop(stack).toComplexDoubleList();
+          c10::complex<double> sum = 0.0;
+          for (int i = 0; i < l.size(); i++) {
```
Consider addressing the `clang-tidy` warning here.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -924,13 +924,20 @@ RegisterOperators reg2({
         std::log(a) / std::log(b),
-        float),
+        float,
+        ".Scalar_Scalar"),
```
`DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION` is called only ever called in this file with `.Scalar_Scalar` as the final arg; consider adding another macro that accepts the first five arguments and then calls `DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION` with `.Scalar_Scalar` as the sixth.

SplitInfinity reviewed (commented):
Just some nits and one or two questions.

anjali411 commented on torch/csrc/jit/runtime/register_prim_ops.cpp:
```diff
@@ -779,8 +779,24 @@ RegisterOperators reg(
+         aten::polar,
+         c10::polar(static_cast<double>(a), static_cast<double>(b)),
+         complex),
+     DEFINE_SCALAR_BINARY_OP_AVOID_COLLISION(
+         aten::polar,
+         c10::polar(static_cast<double>(a), static_cast<double>(b)),
+         c10::polar(static_cast<double>(a), static_cast<double>(b)),
+         Scalar,
+         ".Scalar_Scalar"),
```
I don't think that's true. `c10::complex` calls into `std::complex` (on cpu) which also takes as input absolute value and angle (both of which are real values) https://en.cppreference.com/w/cpp/numeric/complex/polar

SplitInfinity replied:
By discrepancy, I meant discrepancy between the `aten::polar` ops you registered, and `aten::polar` from `native_functions.yaml`.

anjali411 commented on test/jit/test_complex.py:
```diff
@@ -137,3 +172,23 @@ def forward(self, infj: int, nanj: int):
         self.assertEqual(loaded(2, 3), 2 + cmath.infj)
         self.assertEqual(loaded(3, 4), 4 + cmath.nanj)
+
+    def test_comparison_ops(self):
+        def fn1(a: complex, b: complex):
+            return a == b
+
+        def fn2(a: complex, b: complex):
+            return a != b
```
true will add tests for that!

anjali411 commented on test/jit/test_complex.py:
```diff
@@ -111,6 +118,35 @@ def fn(x: complex):
+        def pow_float_complex(x: float, y: complex):
+            return pow(x, y)
+
+        for x, y in zip(vals, complex_vals):
```
good point! 

TODO: update this to `for (x, y) in list(itertools.product(vals, complex_vals)):`

anjali411 committed:
Update on "[TS] Add complex support for more ops"


Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -173,18 +170,148 @@ def forward(self, infj: int, nanj: int):
+        def fn6(a: complex, b: int):
+            return a != b
+
+        with self.assertRaises(RuntimeError):
+            self.checkScript(fn5, (x1, 1))
+        with self.assertRaises(RuntimeError):
+            self.checkScript(fn6, (x1, 1))
```
Can you use `assertRaisesRegex` or `assertRaisesRegexWithHighlight` here instead? That way we can test the error message too.

SplitInfinity commented on torch/csrc/jit/runtime/register_prim_ops_fulljit.cpp:
```diff
@@ -1082,6 +1100,17 @@ RegisterOperators reg2({
         },
         aliasAnalysisFromSchema()),
+    Operator(
+        "aten::sum.complex(complex[] self) -> complex",
```
Is there a test for this? I saw that it was deleted in the latest update.

anjali411 replied:
done!

anjali411 committed:
Update on "[TS] Add complex support for more ops"


Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

[ghstack-poisoned]

anjali411 committed:
Update on "[TS] Add complex support for more ops"


Differential Revision: [D27599114](https://our.internmc.facebook.com/intern/diff/D27599114)

[ghstack-poisoned]

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -299,23 +299,19 @@ def fn4(a: complex, b: float):
-            self.checkScript(fn5, (x1, 1))
-        with self.assertRaises(RuntimeError):
-            self.checkScript(fn6, (x1, 1))
-
```
Why have these tests been removed? I think I commented about them earlier and that's why you added them?

anjali411 replied:
right I saw an `AddressSanitizer` failure on just ASAN after I added them. Just wanted to verify if that was it.

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -99,7 +106,7 @@ def checkCmath(func_name, funcs_template=funcs_template):
+                     'tanh', 'asinh', 'acosh', 'atanh', 'phase', 'isinf', 'isnan', 'isfinite']
 
         # --- Unary ops ---
         for op in unary_ops:
```
One cool thing I learned about recently: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests

Using `with self.subTest(op):` here might make it clear which specific test being generated within this one test is failing.

anjali411 replied:
wow super useful!

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -111,6 +118,32 @@ def fn(x: complex):
+        def pow_float_complex(x: float, y: complex):
+            return pow(x, y)
+
+
```
```suggestion
```

SplitInfinity commented on test/jit/test_complex.py:
```diff
@@ -120,7 +153,6 @@ def func():
         for x in (float_consts + complex_consts):
             checkCmath(x, funcs_template=func_constants_template)
 
-
```
Remove

SplitInfinity reviewed (commented):
I noticed you got rid of tests that we added specifically because we were missing coverage for invalid comparisons between int and complex (or something like that). Can you comment on that?

SplitInfinity reviewed (approved):
Ok approving but I strongly strongly recommend trying to add those tests in a follow up diff. They're not super valuable by themselves but the error logs from the failure make me think that they might have triggered a real issue that we should look into and fix.

anjali411 commented:
@anjali411 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D27599114).
